File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False

        
        stack = [root]
        while len(stack) > 0:
            cur = stack.pop()
            if cur.val == subRoot.val:
                sub_stack = [subRoot]
                copy = [cur]
                while len(sub_stack) > 0 and len(copy) > 0:
                    main = copy.pop()
                    sub = sub_stack.pop()
                    checker = True
                    if sub and not main:
                        checker = False
                        break
                    elif main and not sub:
                        checker = False
                        break
                    if main and sub and main.val != sub.val:
                        checker = False
                        break
                    if main and sub:  
                        copy.append(main.right)
                        copy.append(main.left)
                        sub_stack.append(sub.right)
                        sub_stack.append(sub.left)

                if checker and len(sub_stack) == 0 and len(copy) == 0:
                    return True

            if cur.right: stack.append(cur.right)
            if cur.left: stack.append(cur.left)

        return False
        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (root == nullptr){return nullptr;}
        vector<TreeNode*> que;
        TreeNode *cur, *tmp;
        que.push_back(root);
        while (que.size() > 0)
        {
            cur = que.back();
            que.pop_back();
            tmp = cur->left;
            cur->left = cur->right;
            cur->right = tmp;

            if (cur->right)
            {
                que.push_back(cur->right);
            }
            if (cur->left)
            {
                que.push_back(cur->left);
            }
        }

        return root;
    }
};
